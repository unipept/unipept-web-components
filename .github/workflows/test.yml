name: Test

on: [push]

jobs:
  lint:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - name: Install dependencies
      run: npm install
    - name: Install Jest CLI
      run: npm i -g jest-cli
    - name: Run tests
      run: jest
    
