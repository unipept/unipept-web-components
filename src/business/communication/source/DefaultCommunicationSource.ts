import CommunicationSource from "./CommunicationSource";
import EcResponseCommunicator from "@/business/communication/functional/ec/EcResponseCommunicator";
import GoResponseCommunicator from "./../functional/go/GoResponseCommunicator";
import InterproResponseCommunicator from "./../functional/interpro/InterproResponseCommunicator";
import NcbiResponseCommunicator from "./../taxonomic/ncbi/NcbiResponseCommunicator";
import Pept2DataCommunicator from "./../peptides/Pept2DataCommunicator";

export default class DefaultCommunicationSource implements CommunicationSource {
    public constructor(
        private readonly serviceUrl: string
    ) {}

    public getEcCommunicator(): EcResponseCommunicator {
        return new EcResponseCommunicator();
    }

    public getGoCommunicator(): GoResponseCommunicator {
        return new GoResponseCommunicator();
    }

    public getInterproCommunicator(): InterproResponseCommunicator {
        return new InterproResponseCommunicator();
    }

    public getNcbiCommunicator(): NcbiResponseCommunicator {
        return new NcbiResponseCommunicator();
    }

    public getPept2DataCommunicator(): Pept2DataCommunicator {
        return new Pept2DataCommunicator(this.serviceUrl);
    }
}
