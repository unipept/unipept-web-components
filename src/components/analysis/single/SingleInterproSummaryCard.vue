<template>
    <v-card flat>
        <v-card-text>
            <span v-html="trustLine" class="go-trust"></span>
            <v-select :items="namespaceValues" label="Category" v-model="selectedNamespace"></v-select>
            <amount-table
                annotation-name="InterPro-Entry"
                :item-retriever="getSelectedItemRetriever()"
                :external-url-constructor="getUrl"
                :show-percentage="false"
                :show-namespace="true"
                :rows-per-page="10">
            </amount-table>
        </v-card-text>
    </v-card>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";
import { Prop, Watch } from "vue-property-decorator";
import { Peptide } from "./../../../business/ontology/raw/Peptide";
import FunctionalTrust from "./../../../business/processors/functional/FunctionalTrust";
import { CountTable } from "./../../../business/counts/CountTable";
import InterproDefinition, { InterproCode } from "./../../../business/ontology/functional/interpro/InterproDefinition";
import { InterproNamespace } from "./../../../business/ontology/functional/interpro/InterproNamespace";
import InterproOntologyProcessor from "./../../../business/ontology/functional/interpro/InterproOntologyProcessor";
import { FunctionalUtils } from "./../functional/FunctionalUtils";
import InterproProteinCountTableProcessor
    from "./../../../business/processors/functional/interpro/InterproProteinCountTableProcessor";
import CommunicationSource from "./../../../business/communication/source/CommunicationSource";
import AmountTableItemRetriever from "@/components/tables/AmountTableItemRetriever";
import AmountTable from "@/components/tables/AmountTable.vue";
import SingleAmountTableItemRetriever from "@/components/analysis/single/SingleAmountTableItemRetriever";
import { Ontology } from "@/business";

@Component({
    components: { AmountTable }
})
export default class SingleInterproSummaryCard extends Vue {
    // @ts-ignore
    private namespaceValues: string[] = ["all"].concat(
        Object.values(InterproNamespace).map(x => x.toString()).sort()
    );
    private selectedNamespace: string = "all";

    get interproProteinProcessor(): InterproProteinCountTableProcessor {
        return this.$store.getters.peptideStatus.interproProteinCountTableProcessor;
    }

    get trust(): FunctionalTrust {
        return this.interproProteinProcessor.getTrust();
    }

    get trustLine(): string {
        return FunctionalUtils.computeTrustLine(this.trust, "InterPro-entry", "protein");
    }

    get interproOntology(): Ontology<InterproCode, InterproDefinition> {
        return this.$store.getters.peptideStatus.interproOntology;
    }

    get computedItems(): {
        itemRetriever: AmountTableItemRetriever<InterproCode, InterproDefinition>;
        namespace: string
    }[] {
        const result = [];

        for (const [i, ns] of this.namespaceValues.entries()) {
            let countTable: CountTable<InterproCode>;

            if (ns === "all") {
                countTable = this.interproProteinProcessor.getCountTable();
            } else {
                countTable = this.interproProteinProcessor.getCountTable(ns as InterproNamespace);
            }

            const itemRetriever = new SingleAmountTableItemRetriever(
                countTable,
                this.interproOntology,
                this.trust.totalAmountOfItems
            );

            result.push({
                itemRetriever,
                namespace: ns
            });
        }

        return result;
    }

    private getSelectedItemRetriever(): AmountTableItemRetriever<InterproCode, InterproDefinition> {
        return this.computedItems.filter(i => i.namespace === this.selectedNamespace)[0].itemRetriever;
    }

    private getUrl(code: string): string {
        return `https://www.ebi.ac.uk/interpro/search/text/${code.substr(4)}/#table`;
    }
}
</script>

<style scoped>

</style>
