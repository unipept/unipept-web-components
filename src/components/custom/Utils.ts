export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
