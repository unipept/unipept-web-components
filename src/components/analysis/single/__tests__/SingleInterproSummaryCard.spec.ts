import Setup from "@/test/Setup";
import Vue from "vue"
import Vuetify from "vuetify"
import { createLocalVue, mount } from "@vue/test-utils";
import { waitForCondition, waitForElement } from "@/test/Utils";
import SingleInterproSummaryCard from "@/components/analysis/single/SingleInterproSummaryCard.vue";
import DefaultCommunicationSource from "@/business/communication/source/DefaultCommunicationSource";

Vue.use(Vuetify);

const localVue = createLocalVue();

describe("SingleInterproSummaryCard", () => {
    let vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
    });

    beforeAll(() => {
        const setup = new Setup();
        setup.setupAll();
    });

    it("correctly displays the correct InterPro-entries for a given peptide", async() => {
        const wrapper = mount(SingleInterproSummaryCard, {
            localVue,
            vuetify,
            propsData: {
                peptide: "AALTER",
                equateIl: true,
                communicationSource: new DefaultCommunicationSource()
            }
        });

        // Wait for the table to be rendered.
        await waitForCondition(() => wrapper.findAll("tr").length > 1);

        // We sort the proteins by code, to make the test deterministic
        const sortButton = wrapper.findAll("th").filter(w => w.html().includes("Interpro entry")).at(0);
        sortButton.trigger("click");

        // Wait for this element to be the first in the table, which means sorting is done
        await waitForCondition(() => wrapper.find("tbody tr").html().includes("IPR042201"));

        // We know that the following IPR-entries are associated with the AALTER peptide, and should thus be present.
        const trs = wrapper.findAll("tbody tr");
        expect(trs.at(0).html().includes("IPR000225")).toBeTruthy();
        expect(trs.at(1).html().includes("IPR003613")).toBeTruthy();
        expect(trs.at(2).html().includes("IPR009767")).toBeTruthy();

        // Make sure the total count of items is correct.
        expect(wrapper.find(".v-data-footer__pagination").text()).toEqual("1-9 of 9");
    });

    it("displays the correct trust line", async() => {
        const wrapper = mount(SingleInterproSummaryCard, {
            localVue,
            vuetify,
            propsData: {
                peptide: "AALTER",
                equateIl: true,
                communicationSource: new DefaultCommunicationSource()
            }
        });

        await wrapper.vm.$nextTick();

        // Wait for the table to be rendered.
        await waitForCondition(() => wrapper.find(".interpro-trust").exists() && wrapper.find(".interpro-trust").html().includes("have at least"));

        expect(wrapper.find(".interpro-trust").html()).toMatchSnapshot();
    });
});
