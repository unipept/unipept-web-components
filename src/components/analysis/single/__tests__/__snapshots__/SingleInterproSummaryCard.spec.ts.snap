// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleInterproSummaryCard displays the correct trust line 1`] = `"<span class=\\"interpro-trust\\"><strong>All proteins</strong> have at least one InterPro-entry assigned to them. </span>"`;
